torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp

```diff
@@ -0,0 +1,350 @@
+#include <torch/csrc/jit/passes/fixup_trace_scope_blocks.h>
+
+#include <torch/csrc/jit/passes/canonicalize.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/inliner.h>
+#include <torch/csrc/jit/passes/lower_tuples.h>
+#include <torch/csrc/jit/script/schema_matching.h>
+
+#include <algorithm>
+
+namespace torch {
+namespace jit {
+
+using BlockInfoMap = std::unordered_map<Block*, int>;
+
+void inspectBlocks(Block* b, BlockInfoMap& block_info, int level = 0) {
+  TORCH_INTERNAL_ASSERT(!block_info.count(b));
+  block_info[b] = level;
+  for (Node* n : b->nodes()) {
+    if (n->kind() == prim::FakeScopeBlock) {
+      inspectBlocks(n->blocks()[0], block_info, level + 1);
+    }
+  }
+}
+
+struct RemappedValueInfo {
+  Value* v;
+  int level;
+};
+using RemappingTable = std::unordered_map<Value*, RemappedValueInfo>;
+
+void makeDefsDominateUses(
+    Block* b,
+    RemappingTable& remap,
+    const BlockInfoMap& block_info) {
+  auto find_common_ancestor_block_level = [&](Node* n) {
+    std::unordered_set<Block*> seen;
+
+    Block* my_provenance = b;
+    while (my_provenance) {
+      seen.insert(my_provenance);
+      if (!my_provenance->owningNode()) {
+        my_provenance = nullptr;
+        continue;
+      }
+      my_provenance = my_provenance->owningNode()->owningBlock();
+    }
+
+    Block* their_provenance = n->owningBlock();
+    while (their_provenance) {
+      if (seen.count(their_provenance)) {
+        return block_info.at(their_provenance);
+      }
+      if (!their_provenance->owningNode()) {
+        their_provenance = nullptr;
+        continue;
+      }
+      their_provenance = their_provenance->owningNode()->owningBlock();
+    }
+    TORCH_INTERNAL_ASSERT(false);
+  };
+
+  auto process_node = [&](Node* n) {
+    // inspect uses
+    for (size_t i = 0; i < n->inputs().size(); ++i) {
+      Value* inp = n->inputs()[i];
+      if (inp->node()->owningBlock() != b) {
+        int ancestor_level = find_common_ancestor_block_level(inp->node());
+
+        Block* origin_block = inp->node()->owningBlock();
+        if (ancestor_level == block_info.at(origin_block)) {
```

Comment:
I don't think this special case needs to exist. It seems like it is just the degenerate version of the code below.